driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -308,139 +309,66 @@ public void sendAndReceiveAsync(final CommandMessage message, final SingleResult
 
         try {
             message.encode(bsonOutput);
-            String commandName = sendStartedEvent(bsonOutput, message);
+            CommandEventSender commandEventSender = new CommandEventSender(message);
+            commandEventSender.sendStartedEvent(bsonOutput);
 
-            long startTimeNanos = System.nanoTime();
-            if (sendCompressor == null || SECURITY_SENSITIVE_COMMANDS.contains(commandName)) {
-                sendCommandMessageAsync(message, callback, bsonOutput, commandName, startTimeNanos);
+            if (sendCompressor == null || SECURITY_SENSITIVE_COMMANDS.contains(commandEventSender.getCommandName())) {
+                sendCommandMessageAsync(message.getId(), decoder, callback, bsonOutput, commandEventSender);
             } else {
-                CompressedMessage compressedMessage = new CompressedMessage(RequestMessage.OpCode.OP_QUERY, bsonOutput.getByteBuffers(),
+                CompressedMessage compressedMessage = new CompressedMessage(message.getOpCode(), bsonOutput.getByteBuffers(),
                                                                                    sendCompressor, getMessageSettings(description));
                 compressedMessage.encode(compressedBsonOutput);
                 bsonOutput.close();
-                sendCommandMessageAsync(message, callback, compressedBsonOutput, commandName, startTimeNanos);
+                sendCommandMessageAsync(message.getId(), decoder, callback, compressedBsonOutput, commandEventSender);
             }
         } catch (RuntimeException e) {
```

Comment:
Convert to `Throwable localThrowable` ?

Could either `bsonOutput.close()` or `compressedBsonOutput.close()` throw an exception? If so the callback wouldnt be called.